### PR TITLE
Add pruning (lottery tickets)

### PR DESCRIPTION
**Pull request by avishreekh**
Solves #50

### CODE REVIEW

File: logs.rst [hunk 1]
```diff
@@ -8,3 +8,14 @@ Results while adding implementations -
...
+Results for Pruning using Lottery Tickets - 
+===========================
+
++-----------------------------------------------------------+----------------------------------+----------------------+---------------------+
+|  Model used                                               |  Pruning percent                 | Epochs               | Test accuracy(ies)  |
++===========================================================+==================================+======================+=====================+
+|  ResNet18 (on MNIST)                                      |  0%                              |                      |     0.9829          |
+|                                                           |  10%                             |         3            |     0.9816          |
+|                                                           |  19%                             |                      |     0.9851          |
++-----------------------------------------------------------+----------------------------------+----------------------+---------------------+
```
Comment (Het-Shah) [inline logs.rst:21]: Can you make a column for dataset separately, and please do it for DML as well. Even that is trained on MNIST as far as I remember.

Also, sorry for not mentioning before, train it for 5 epochs to maintain uniformity. 

cc: @NeelayS

File: KD_Lib/Pruning/lottery_tickets/lottery_tickets.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+            total_params = np.prod(tensor.shape)
+            nonzero += nz_count
+            total += total_params
+            print(
+                f"{name:20} | nonzeros = {nz_count:7} / {total_params:7} ({100 * nz_count / total_params:6.2f}%) | total_pruned = {total_params - nz_count :7} | shape = {tensor.shape}"
+            )
+        print(
+            f"alive: {nonzero}, pruned : {total - nonzero}, total: {total}, Compression rate : {total/nonzero:10.2f}x  ({100 * (total-nonzero) / total:6.2f}% pruned)"
+        )
+        return round((nonzero / total) * 100, 1)
```
Comment (Het-Shah) [inline KD_Lib/Pruning/lottery_tickets/lottery_tickets.py:204]: I guess do one thing. If `model_path` is `None` then print statistics for all the iterations 
(Of what I think from this code it is printing only for the one model if it it None)

File: KD_Lib/Pruning/lottery_tickets/lottery_tickets.py [hunk 3]
```diff
@@ -97,6 +97,7 @@ def _original_initialization(self):
 
     def _train_after_pruning(self, prune_it):
         best_acc = 0.0
+        best_weights = None
```
Comment (Het-Shah) [inline KD_Lib/Pruning/lottery_tickets/lottery_tickets.py:100]: initialize this to model weights (the reason being if the accuracy doesn't improve then weights will be None)

Comment (avishreekh) [reply KD_Lib/Pruning/lottery_tickets/lottery_tickets.py:100]: I have already included a condition while saving the model that if best_weights are None, then nothing needs to be done

File: KD_Lib/Pruning/lottery_tickets/lottery_tickets.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    :param model (torch.nn.Module): Model that needs to be pruned
+    :param train_loader (torch.utils.data.DataLoader): Dataloader for training
+    :param test_loader (torch.utils.data.DataLoader): Dataloader for validation/testing
+    :param loss_fn (torch.nn.Module): Loss function to be used for training
+    """
+
+    def __init__(self, model, train_loader, test_loader, loss_fn=nn.CrossEntropyLoss()):
+
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        self.model = model.to(self.device)
```
Comment (Het-Shah) [inline KD_Lib/Pruning/lottery_tickets/lottery_tickets.py:20]: take device as an input in `__init__`
